### PR TITLE
V3 service instances can be filtered by space guid

### PR DESCRIPTION
**Pull request by Samze**
As an app dev I can filter v3 service instances by space guid [#152922808](https://www.pivotaltracker.com/story/show/152922808)

## What

This PR adds a space relationship to v3 service instances. This is required by the CF cli to complete the share-service command.

Changes:
* Alter presenter to add space relationship & link
* GET /v3/service_instances allows filtering by space_guid
* Updated docs & added a service_instance object description

## Notes
* The space contained in the relationship is the space in which the service instance was created.

## PR 

* [X] I have viewed signed and have submitted the Contributor License Agreement
* [X] I have made this pull request to the `master` branch
* [X] I have run all the unit tests using `bundle exec rake`
* [X] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats) on bosh lite

Thanks, sapi (@samze and @deniseyu)

### CODE REVIEW

File: spec/unit/controllers/v3/service_instances_controller_spec.rb [hunk 1]
```diff
@@ -23,6 +23,28 @@
...
+            expect(parsed_body['resources'].length).to eq 2
+
+            response_names = parsed_body['resources'].map { |resource| resource['name'] }
+            expect(response_names).to include(service_instance.name, service_instance2.name)
+          end
+        end
+
+        context 'when service instances are filtered by space guid' do
+          it 'returns only the matching service instances' do
+            get :index, space_guids: 'space-2-guid, space-3-guid'
```
Comment (Gerg) [inline spec/unit/controllers/v3/service_instances_controller_spec.rb]: There is an extra space between `space-2-guid` and `space-3-guid`

Comment (Gerg) [reply spec/unit/controllers/v3/service_instances_controller_spec.rb]: Or, is that intentional? Do we test cases like this elsewhere?

File: spec/unit/messages/service_instances_list_message_spec.rb [hunk 2]
```diff
@@ -56,5 +60,23 @@ module VCAP::CloudController
...
+      context 'names' do
+        it 'validates names is an array' do
+          message = ServiceInstancesListMessage.new names: 'tricked you, not an array'
+          expect(message).to be_invalid
+          expect(message.errors[:names]).to include('must be an array')
+        end
+      end
+
+      context 'space guids' do
+        it 'validates app_guids is an array' do
```
Comment (Gerg) [inline spec/unit/messages/service_instances_list_message_spec.rb]: space_guids

File: spec/unit/queries/service_instance_list_fetcher_spec.rb [hunk 3]
```diff
@@ -19,18 +20,43 @@ module VCAP::CloudController
...
           it 'only returns matching service instances' do
             results = fetcher.fetch_all(message: message).all
-            expect(results).to match_array([service_instance_1])
-            expect(results).not_to include(service_instance_2)
+            expect(results).to match_array([service_instance_1, service_instance_2])
+            expect(results).not_to include(service_instance_3)
+          end
+        end
+
+        context 'by space guid' do
```
Comment (Gerg) [inline spec/unit/queries/service_instance_list_fetcher_spec.rb]: The query corresponding to this is a bit complicated, so I'm having trouble reasoning about it. Could you add a couple more defensive tests at least covering the following:
- Filtering by a space guid that has no service instances
- Filtering by an empty list of space guids

File: spec/unit/controllers/v3/service_instances_controller_spec.rb (hunk 1 above)
Comment (deniseyu) [reply spec/unit/controllers/v3/service_instances_controller_spec.rb]: We wanted to test that the parsing logic ignores whitespaces at some point, but this probably isn't the right place for that. We'll change it to make it consistent.

File: spec/unit/messages/service_instances_list_message_spec.rb (hunk 2 above)
Comment (deniseyu) [reply spec/unit/messages/service_instances_list_message_spec.rb]: Oops, fixed.

File: spec/unit/queries/service_instance_list_fetcher_spec.rb (hunk 3 above)
Comment (deniseyu) [reply spec/unit/queries/service_instance_list_fetcher_spec.rb]: No problem - thanks for spotting this. We've also added a few more edge case tests.

File: spec/unit/controllers/v3/service_instances_controller_spec.rb (hunk 1 above)
Comment (Samze) [reply spec/unit/controllers/v3/service_instances_controller_spec.rb]: Changes for this included in #1033 and can be viewed here: https://github.com/cloudfoundry-incubator/cloud_controller_ng_sapi/commit/ff8276e5d201e573df03640a0f6865288b2d55af#diff-eb9dfb65b6b62e5646cab0d740db45e3R41

File: spec/unit/messages/service_instances_list_message_spec.rb (hunk 2 above)
Comment (Samze) [reply spec/unit/messages/service_instances_list_message_spec.rb]: Changes for this included in #1033 and can be viewed here: https://github.com/cloudfoundry-incubator/cloud_controller_ng_sapi/commit/ff8276e5d201e573df03640a0f6865288b2d55af#diff-58747ec5e09fb7b688e999c46413b076R74

File: spec/unit/queries/service_instance_list_fetcher_spec.rb (hunk 3 above)
Comment (Samze) [reply spec/unit/queries/service_instance_list_fetcher_spec.rb]: Changes for this included in #1033 and can be viewed here: https://github.com/cloudfoundry-incubator/cloud_controller_ng_sapi/commit/ff8276e5d201e573df03640a0f6865288b2d55af#diff-212dc751d0b344666f773f5f85522b5cR48

File: app/presenters/v3/service_instance_presenter.rb [hunk 4]
```diff
@@ -9,12 +9,28 @@ def to_hash
...
-            name:      service_instance.name
+            name:      service_instance.name,
+            relationships: {
+              space: {
+                data: {
+                  guid: service_instance.space.guid
+                }
+              }
+            },
+            links: {
```
Comment (elenasharma) [inline app/presenters/v3/service_instance_presenter.rb]: To construct `links` in our presenters, we use a `build_links` method, as in [organization_presenter](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/app/presenters/v3/organization_presenter.rb#L19-L25).

We'll merge this PR, but in a follow-up PR could you change this presenter to follow this convention?
